### PR TITLE
Ruby declared license scan

### PR DESCRIPTION
**Pull request by csasarak**
# Overview

This PR adds a parser that looks for assignment statements in ruby gemspec files and parses out the licenses. 

## Acceptance criteria

The goal is to provide a way for fossa-cli to analyze bundler projects and determine what license it has. Gemspec files are ruby code, but usually licenses are written in as static text. The reason to prefer parsing ourselves was to avoid needing to run the ruby code in the gemspec file in order to determine the declared license. 

## Testing plan

Unit tests test both the parser layer as well as the license discovery layer.

I wrote a script to pull down some of the most popular ruby gems and got about 60. I tested pathfinder against all 60 and determined that the detected licenses were correct in each of them, aside from 3 which weren't bundler projects at all.

## Risks

Please pay special attention to the tests and let me know if there are common scenarios I may have missed when writing this.

## References
Closes fossas/team-analysis#565

## Checklist

- [x] I added tests for this PR's change (or confirmed tests are not viable).
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] I updated `Changelog.md` if this change is externally facing. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] I updated `*schema.json` if I have made changes for `.fossa.yml`, `fossa-deps.{json, yaml, yml}`. You may also need to update these if you have added/removed new dependency (e.g. pip) or analysis target type (e.g. poetry).
- [x] I linked this PR to any referenced GitHub issues, if they exist.

### CODE REVIEW

File: src/Strategy/Ruby/Gemspec.hs
Comment (meghfossa) [inline src/Strategy/Ruby/Gemspec.hs:46]: [nit] would like to see type signature in where block.

File: src/Strategy/Ruby/Gemspec.hs
Comment (meghfossa) [inline src/Strategy/Ruby/Gemspec.hs:32]: [nit] Would like to see comment with example text it ought to be parsing (few cases). Here is good [example](https://github.com/fossas/fossa-cli/blob/master/src/App/Fossa/VSI/DynLinked/Internal/Binary.hs#L108). 

I'm guilty of not doing this too, but given this parser seems bit involved, example case and edge case illustrated would help in reviewing. 

This applies to other applicable parser as well in this PR.

File: src/Strategy/Ruby/Gemspec.hs
Comment (spatten) [inline src/Strategy/Ruby/Gemspec.hs:40]: I'm not familiar with how this parsing library works, but does this deal with escaped delimiters within the string?

E.g. `"This is a \"string\" with quotes in it"` or `'we don\'t need to worry about escapes'`

There are similar worries for arbitrary delimiters: `%Q{this string \{keeps on going\} until here}`

This isn't too likely for the data-set we're looking at here (hopefully SPDX expressions), but if it's easy to deal with it'd be nice.

Comment (spatten) [reply src/Strategy/Ruby/Gemspec.hs:40]: Also, for strings delimited by `"` or strings with arbitrary delimiters that start with `%Q`, you can have string interpolation like this:

```
"hello, #{name}"
%Q{hello, #{name}}
```

Should we just fail if we see string interpolation?

File: src/Strategy/Ruby/Gemspec.hs
Comment (spatten) [inline src/Strategy/Ruby/Gemspec.hs:37]: It's idiomatic to say `.freeze`, but `.freeze()` is also valid Ruby code. Should we deal with that too?

Comment (spatten) [reply src/Strategy/Ruby/Gemspec.hs:37]: This might be going too far, but we could probably just ignore anything after a `.` following a closing string delimiter.

This would deal with things like `license = 'mit'.upcase` or `license = '  Apache 2.0 '.trim`

I'd say we ignore this if you didn't see it when you downloaded 1000 gemspecs or whatever the number was.

File: src/Strategy/Ruby/Gemspec.hs
Comment (csasarak) [reply src/Strategy/Ruby/Gemspec.hs:59]: Hi Megh, I've updated with text around each of the parsers. I hope they make it a little more clear what they do

File: src/Strategy/Ruby/Gemspec.hs
Comment (csasarak) [reply src/Strategy/Ruby/Gemspec.hs]: I'm going to mark this resolved per our earlier conversation.

File: src/Strategy/Ruby/Gemspec.hs
Comment (csasarak) [reply src/Strategy/Ruby/Gemspec.hs:65]: I've added support for escaping the closing delimiter in strings. In all my tests I didn't see this situation but it's good to have.

File: src/Strategy/Ruby/Gemspec.hs
Comment (spatten) [inline src/Strategy/Ruby/Gemspec.hs:132]: Sorry I didn't notice this the first time through, but array literals can also use the same delimiters as string literals. Yet another case where Ruby took too much inspiration from Perl.

https://en.wikibooks.org/wiki/Ruby_Programming/Syntax/Literals#The_.25_Notation

So these all result in an array of words:

```ruby
a = %w(one two three)
a = %w{one two three}
a = %w#one two three#
a = %w<one two three>
```

The exact same thing happens for the `W` variants, but with a capital W you can put string interpolation inside of the words (which we should ignore like we are ignoring it for strings).

I think this might be worth fixing, but only because we already have the machinery you already built for strings that will make this reasonably easy to do.

File: src/Strategy/Bundler.hs
Comment (spatten) [inline src/Strategy/Bundler.hs:96]: Did you ever run into this case from that URL?

> You can also use a custom license file along with your gemspec and specify a LicenseRef-<idstring>, where idstring is the name of the file containing the license text.

I had no idea that that existed. I'm mostly asking for curiousity's sake — unless you saw this in the one of the gemspecs you downloaded, I think we should fix this if and when we see it in the wild.

Comment (csasarak) [reply src/Strategy/Bundler.hs:96]: I never encountered that and totally missed it myself. For now I think I'll hold off on implementing it because it seems kind of arcane.

File: src/Strategy/Ruby/Gemspec.hs
Comment (csasarak) [reply src/Strategy/Ruby/Gemspec.hs]: I have incorporated this into my latest commit.

**Review by spatten — APPROVED** [review]
This is looking good to me from a "how Ruby works" side of things.

I think we'll also want an approval from @meghfossa on the Haskell side of things before merging

**Review by meghfossa — APPROVED** [review]

**Pull request merged**
